experimental/lib/Transforms/ResourceSharing/Crush.cpp

```diff
@@ -0,0 +1,682 @@
+        unionGroup.insert(unionGroup.end(), g2->begin(), g2->end());
+        sharingGroups.push_back(unionGroup);
+        sharingGroups.erase(g1);
+        sharingGroups.erase(g2);
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+void logGroups(const SharingGroups &sharingGroups, NameAnalysis &namer) {
+  for (const auto &group : sharingGroups) {
+    llvm::errs() << "group : {";
+    for (auto *op : group) {
+      llvm::errs() << namer.getName(op) << " ";
+    }
+    llvm::errs() << "}\n";
+  }
+}
+
+void sortGroups(SharingGroups &sharingGroups, FuncPerfInfo &info) {
+  for (auto &g : sharingGroups) {
+    // use bubble sort to sort each group:
+    if (g.size() <= 1)
+      continue;
+    bool modified = false;
+
+    do {
+      modified = false;
+      for (size_t i = 1; i < g.size(); i++) {
+        for (auto cf : info.critCfcs) {
+
+          auto op1 = info.cfSccs[cf].find(g[i - 1]);
+          auto op2 = info.cfSccs[cf].find(g[i]);
+          if (op1 != info.cfSccs[cf].end() && op2 != info.cfSccs[cf].end() &&
+              op1->second > op2->second) {
+            iter_swap(g.begin() + i, g.begin() + i - 1);
+            modified = true;
+          }
+        }
+      }
+    } while (modified);
+  }
+}
+
+// Set opOccupancy[op] to the occupancy required to achieve maximum performance
+// of all performance critical CFCs.
+void getOpOccupancy(const SmallVector<Operation *> &sharingTargets,
+                    llvm::MapVector<Operation *, double> &opOccupancy,
+                    TimingDatabase &timingDB, FuncPerfInfo &funcPerfInfo) {
+
+  double latency;
+  for (auto *target : sharingTargets) {
+    // By default, the op is assigned with no occupancy. If a performance
+    // critical CFC contains that op, then we set the occupancy to the occupancy
+    // of op in that CFC.
+    opOccupancy[target] = 0.0;
+    for (auto cf : funcPerfInfo.critCfcs) {
+      if (funcPerfInfo.cfUnits[cf].find(target) !=
+          funcPerfInfo.cfUnits[cf].end()) {
+        if (failed(timingDB.getLatency(target, SignalType::DATA, latency)))
+          latency = 0.0;
+        // Occupancy = Latency / II = Latency * Throughput.
+        opOccupancy[target] = latency * funcPerfInfo.cfThroughput[cf];
+      }
+    }
+  }
+}
+
+LogicalResult
+sharingWrapperInsertion(handshake::FuncOp &funcOp, SharingGroups &sharingGroups,
+                        MLIRContext *ctx,
+                        MapVector<Operation *, double> &opOccupancy) {
+  OpBuilder builder(ctx);
+  for (auto group : sharingGroups) {
+
+    // If the group only has one operation or has no operations, then go to next
+    // group
+    if (group.size() <= 1)
+      continue;
+
+    // Elect one operation as the shared operation.
+    auto *sharedOp = *group.begin();
+
+    // Enumerate the operands of the original pre-sharing operations.
+    llvm::SmallVector<Value> sharingWrapperInputs;
+    for (auto *op : group) {
+      for (auto val : op->getOperands()) {
```

Comment:
`llvm::copy(op->getOperands(), std::back_inserter(sharingWrapperInputs))` is more standard here.